src/main.rs

```diff
@@ -65,9 +67,112 @@ pub struct Config {
+        // This line of code takes what might be a relative path and returns an absolute path
+        // without any `.` or `..` in. If the path then points to a symbolic link, the code then
+        // follows the link, repeating until it finds something which is not a symbolic link. This
+        // is what is set as the `root_dir`.
+        //
+        // Doing this makes it possible to report the real root directory in the log, and also
+        // makes checking that a file is actually in the root directory more robust.
+        let canonical_root_dir = self.root_dir.canonicalize()?;
+        self.root_dir = canonical_root_dir;
+        Ok(())
+    }
+
+    /// Checks if the given path is in the root dir.
+    ///
+    /// If it is, return its canonical representation. If it isn't, return an error. This function
+    /// will error if the path does not point to an actual file or directory.
+    fn check_in_root_dir(&self, path: PathBuf) -> Result<PathBuf> {
+        let path = path.canonicalize()?;
+
+        // Skip the check if we've configured to allow files outside the root.
+        if self.allow_escape_root || path.starts_with(&self.root_dir) {
+            Ok(path)
+        } else {
+            return Err(Error::EntityNotInRoot);
+        }
+    }
+
+    /// Check if the request has the required password (if we set one).
+    fn check_auth(&self, req: &Request<Body>) -> bool {
+        // This macro avoids us having to write out a match for every step.
+        macro_rules! err_to_ret {
+            ($e:expr) => {
+                match $e {
+                    Ok(v) => v,
+                    Err(_) => return false,
+                }
+            };
+        }
+
+        let reference_auth = match self.auth.as_ref() {
+            Some(auth) => auth,
+            // If there is no password, carry on serving the request.
+            None => return true,
+        };
+
+        // Get and decode the auth token
+        let headers = req.headers();
+        let auth_header = match headers.get(header::AUTHORIZATION) {
+            Some(header) => header,
+            // If the header isn't set, then send a request for auth.
+            None => return false,
+        };
+        let auth_header = err_to_ret!(auth_header.to_str());
+
+        if !matches!(auth_header.get(..6), Some(s) if s.eq_ignore_ascii_case("basic ")) {
+            return false;
+        }
+        let auth = match auth_header.get(6..).map(|s| s.trim()) {
+            Some(auth) => auth,
+            None => return false,
+        };
+        let auth = err_to_ret!(base64::decode(auth));
+        let auth = err_to_ret!(str::from_utf8(&auth));
+        let auth: Auth = err_to_ret!(auth.parse());
+        *reference_auth == auth
+    }
+}
+
+#[derive(Clone, PartialEq)]
+struct Auth {
+    username: String,
+    password: String,
+}
+
+impl std::str::FromStr for Auth {
+    type Err = &'static str;
+    fn from_str(s: &str) -> std::result::Result<Self, Self::Err> {
```

Comment:
I've made this parser very lenient, to the point that it cannot fail! I think it's still a good idea to leave it in `FromStr`, so it could be made less lenient in the future by only altering code here.